pkg/parse/updater.go

```diff
@@ -186,21 +186,9 @@ func (u *Updater) update(ctx context.Context, cache *cacheForCommit) status.Mult
 		}
 	}
 
-	// Apply the declared resources
-	if !cache.applied {
-		declaredObjs, _ := u.Resources.DeclaredObjects()
-		_, err := u.apply(ctx, declaredObjs, cache.source.commit)
-		if err != nil {
-			return err
-		}
-		// Only mark the commit as applied if there were no (non-blocking) parse errors.
-		// This ensures the apply will be retried until parsing fully succeeds.
-		if cache.parserErrs == nil {
-			cache.applied = true
-		}
-	}
-
 	// Update the resource watches (triggers for the Remediator).
+	// Start watches before the applier runs to detect any drift that may occur
```

Comment:
Ah looks like this fails for CRDs if a CRD and CR are added in the same commit, because the watch can't be started for a CRD that does not exist yet. @nan-yu do you have any thoughts for a workaround?

Reply:
Hmm, it seems we have to update watches in two steps. Before apply, we update watches with known GVKs, and update watches again for those new registered GVKs.

Reply:
I'm wondering are we overthinking this issue? Even though  `u.apply` and `u.watch` is not atomic, the possibility that an applied object being mutated before the watch is running seems very low.

Regarding the prune issue and CRD/CR in the same commit issue, I'm okay with apply first and then update watches.

Reply:
The mitigation for this is a lot more complicated than you might think, especially because the mapper needs resetting when new CRDs are added or removed.

For cli-utils, I wrote the DefaultStatusWatcher & ObjectStatusReporter to handle this:
- https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/kstatus/watcher/object_status_reporter.go#L47
- https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/kstatus/watcher/default_status_watcher.go

We could use the DefaultStatusWatcher/ObjectStatusReporter here, but it might require significant code changes. And it would duplicate all the watches being made by the applier.

Reply:
FWIW, the duplication of watches was one of the primary drivers for my suggesting merging the remediator into cli-utils applier and making it do continuous reconciliation, instead of just periodic applies. But that project is pretty significant and hasn't been prioritized.

Reply:
So looking deeper into this, there's a few problems:

1. If the watch [fails to start](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/pkg/remediator/watch/filteredwatcher.go#L387) any calls to [Stop will result in a deadlock](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/pkg/remediator/watch/filteredwatcher.go#L220), because `start` doesn't let go of the lock until the watch successfully starts. This can be worked around relatively simply by passing in a cancellable context into `startWatch` and calling `cancel` in `Stop`.
2. Errors from starting the watch are [not returned synchronously](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/pkg/remediator/watch/manager.go#L206), so we don't actually know if any of the watches successfully started after calling `UpdateWatches`. This means we don't actually have a guarantee that the remediator watches are running before we invoke the applier. Furthermore [the watcher may fail and restart watches](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/pkg/remediator/watch/filteredwatcher.go#L298) at any point, which means that we could be missing watch events during watcher restarts.

[1] is worth fixing regardless IMO, however [2] adds quite a bit of complexity to this problem.

Reply:
Regarding [2], does passing in the [resourceVersion](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/pkg/remediator/watch/filteredwatcher.go#L298) make it so that we capture events between watch restarts?